File: clojure/nucleotide-count/src/nucleotide_count.clj
(ns nucleotide-count
  (:require
    [clojure.string :as str]))

(defn dna-alphabet [] (list \A \C \G \T))

(defn count-of-nucleotide-in-strand [nucleotide strand]
  (if (not (some #(= nucleotide %) (dna-alphabet)))
    (throw (Exception. (format "Nucleotide %s does not exist" nucleotide)))
    (if (str/blank? strand)
      0
      (count (filterv #(= (str nucleotide) (str %)) (re-seq #"." strand))))))

(defn nucleotide-counts [strand]
  (if (str/blank? strand)
    (into {} (map #(hash-map % 0) (dna-alphabet)))
    (into {} (map #(hash-map % (count-of-nucleotide-in-strand % strand)) (dna-alphabet)))))

File: lua/nucleotide-count/nucleotide-count.lua
local DNA = {}

local function parse (s)
  local counts = { A = 0, T = 0, C = 0, G = 0 }
  for i=1,#s do
    local nucleotide = s:sub(i,i)
    if not nucleotide:match("[ATCG]+") then error"Invalid Nucleotide" end
    counts[nucleotide] = tonumber(counts[nucleotide]) + 1
  end
  return counts
end

function DNA:new (strand)
  local o = {}
  setmetatable(o, self)
  self.__index = self
  self.strand = strand
  self.nucleotideCounts = parse(strand)
  return o
end

function DNA:count (nucleotide)
  if not nucleotide:match("[ATCG]+") then error"Invalid Nucleotide" end
  return self.nucleotideCounts[nucleotide]
end

return DNA
